### PR TITLE
Review from @anorth at 0d7b13a5

### PR DESCRIPTION
**Pull request by anorth**
This PR is to facilitate review of the state of `master` at 0d7b13a5. This is not intended for merge. Instead, apply changes in response to feedback in a new branch from master.

### CODE REVIEW

File: api/actors.yml [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    properties:
+      kind:
+        type: string
+        readOnly: true
+        enum: [actor]
+      id: 
+        type: string
+      role:
+        type: string
+        enum: [init,account,miner,storageMarket,paymentBroker]
```
Comment (anorth) [inline api/actors.yml]: Unless there's a reason, should probably match the capitalization and names of the spec. `Init`, `Account`, `StorageMiner`, `StorageMarket`, `PaymentChannel`, `MultisigAccount`.

File: api/actors.yml [hunk 2]
```diff
@@ -0,0 +1,149 @@
+models:
+  Actor:
+    type: object
+    properties:
+      kind:
+        type: string
+        readOnly: true
+        enum: [actor]
+      id:
```
Comment (anorth) [inline api/actors.yml]: I think this is intended to be its address, and we should use that name. There are multiple schemas for addresses, one of them confusing called "ID address".

File: api/actors.yml [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      role:
+        type: string
+        enum: [init,account,miner,storageMarket,paymentBroker]
+      nonce:
+        type: number
+        description: Nonce is the nonce expected on the next message from this actor. Messages are processed in strict, contiguous nonce order. 
+      balance:
+        type: number
+        description: Balance is the amount of FIL in the actor's account.
+      stateRoot:
```
Comment (anorth) [inline api/actors.yml]: `head` (state root is used elsewhere).

Also include the `codeCid`.

File: api/actors.yml [hunk 4]
```diff
@@ -0,0 +1,149 @@
+models:
+  Actor:
```
Comment (anorth) [inline api/actors.yml:2]: While it's bound to rot, it's still probably worthwhile to link to the spec for spec-related things, with an explicit disclaimer that these structures are not intended to match the spec exactly (indeed, they are intended to be more usable), but should generally agree with it.

https://github.com/filecoin-project/specs/blob/master/data-structures.md#actor

File: api/actors.yml [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  EmptyActorState:
+    type: object 
+    description: There is no additional state exposed by this actor
+  InitActorState:
+    type: object
+    properties:
+      nextId: 
+        type: number
+        description: The next numerical ID to allocate to a newly instantiated Actor
+  MinerActorState:
```
Comment (anorth) [inline api/actors.yml]: `StorageMinerActorState`

File: api/actors.yml [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        type: array
+        items: 
+          $ref: 'actors.yml#/models/Sector'
+      slashedAt:
+        type: number
+      ownedStorageCollateral:
+        type: number
+      provingPeriodEnd:
+        type: number
+  PaymentBrokerActorState:
```
Comment (anorth) [inline api/actors.yml]: The payment broker will be replaced by individual payment channel actors.

File: api/actors.yml [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        description: Filecoin's total committed storage as of the current block.
+endpoints:
+  actors_collection:
+    get:
+      operationId: listActors
+      summary: List Actors
+      description: List Actors currently known to this node in ID order
+      tags:
+        - Actors
+      parameters:
```
Comment (anorth) [inline api/actors.yml:117]: This, and all state query operations, must take a tipset key as an optional parameter, indicating the point in the chain history at which to query. Whether or not it's specified, the result should include an envelope that specifies the tipset at which the query was actually evaluated.

File: api/actors.yml [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    properties:
+      totalStorage: 
+        type: number
+        description: Filecoin's total committed storage as of the current block.
+endpoints:
+  actors_collection:
+    get:
+      operationId: listActors
+      summary: List Actors
+      description: List Actors currently known to this node in ID order
```
Comment (anorth) [inline api/actors.yml]: ID -> Address everywhere in this file

File: api/api.yml [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+- url: http://127.0.0.1:3453/api/chain/v0
+security:
+  - 'Bearer Token Authentication': []
+components:
+  securitySchemes:
+    'Bearer Token Authentication':
+      type: http
+      scheme: Bearer Token
+      description: |
+        V0 of this API supports all or nothing authentication and authorization suitable for use by the owner/operator of the current node, and nothing else. Callers of the API must provide a bearer token included in the Authorization header of each request. This token should be provided by the node for use by the caller via some out of band mechanism (like a file created on the filesystem upon initialization or upon startup).
```
Comment (anorth) [inline api/api.yml]: "Callers of the API may provide". Reword this to make it clear that authentication is optional by the server, but if it wants to do bearer authentication, this is the standard way. Implementations could provide other site-specific auth methods too.

File: api/chain.yml [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+              description: A base64 encoded string of the proof of delay during computation of vdfResult
+              example: nkgYSBwZ...
+      electionProof:
+        type: string
+        description: A base64 encoded byte array that represents the proof that the miner of this Block is a leader at this round.
+        example: V8896JC51ikqPVqYUCVbHzUfaOgjTBF6BYc8G/2P9CJMAg8MJzOi1j3F8hJSj/C6tKTLd9qHpLEjRpiGUzz04gE=
+      parentTipsetId:
+        type: string
+        description: The ID of the Tipset on which this block was based. In the spec this field is defined as an array of Block IDs, which is effectively equivalent to a Tipset ID in this API.
+        example: bafy2bzacebnsn4ywt2cgejgp32rgsv4vaylrqe32axw4cfadxhoutxxb7puzs
```
Comment (anorth) [inline api/chain.yml:71]: Give an example that has >1 block

File: api/chain.yml [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        description: A base64 encoded representation of the signature over the hash of this entire block with the miner's worker key to ensure that it is not tampered with after creation
+        example: TODO
+  ExecutedMessage:
+    type: object
+    properties:
+      kind:
+        type: string
+        enum: [executedMessage]
+        readOnly: true
+      id:
```
Comment (anorth) [inline api/chain.yml:99]: The executed message is an abstraction over a message + receipt for API convenience, and doesn't have an ID. The message itself does have a CID though.

I think the implication that you can't fetch an individual executed message is ok.

File: api/chain.yml [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+        example: 1568753711
+      blsAggregate:
+        type: string
+        description: A base64 encoded representation of the aggregated BLS signature for all the messages in this block that were signed using BLS signatures
+        example: TODO
+      blockSig:
+        type: string
+        description: A base64 encoded representation of the signature over the hash of this entire block with the miner's worker key to ensure that it is not tampered with after creation
+        example: TODO
+  ExecutedMessage:
```
Comment (anorth) [inline api/chain.yml:92]: Heads up - this might get more complicated. We might move receipts out of the block, which will make this difficult to construct. See https://github.com/filecoin-project/specs/issues/522

We'll have to think about this, but in the meantime it might be sufficient to remove this abstraction in favour of just exposing resources shaped more like the underlying message and receipt.

File: api/chain.yml [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+    type: object
+    properties:
+      kind: 
+        type: string
+        enum: [tipset]
+        readOnly: true
+      id:
+        type: string
+        readOnly: true
+        example: 'zDPWYqFCyaQ9QJHr1qWZ19rFg3YRkcuXREkvpiHRJZhGN5T8SnyJ'
```
Comment (anorth) [inline api/chain.yml:162]: Show an example with >1 block

File: api/chain.yml [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+      responses:
+        200:
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/models/Message'
+        401:
+          $ref: 'api.yml#/components/responses/UnauthorizedError'
+  messages_collection:
```
Comment (anorth) [inline api/chain.yml:239]: You can add a GET here which returns the list of messages in a message collection (the id of which is in the header)

File: api/chain.yml [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+    post:
+      operationId: createMessage
+      summary: Send a Message
+      description: This method sends a new Message to an Actor recipient.
```
Comment (anorth) [inline api/chain.yml:243]: We'll have to be a bit more explicit here about whether the message is already signed, or is to be signed by the node (and if so, with what key), etc, etc. Ok to pick just one answer for now, but it should be clear.

File: api/chain.yml [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+        401:
+          $ref: 'api.yml#/components/responses/UnauthorizedError'
+  tipsets_collection:
+    get:
+      operationId: listTipsets
+      summary: List Tipsets
+      description: List Tipsets currently known to this node in reverse height order
+      tags:
+        - Tipsets
+      parameters:
```
Comment (anorth) [inline api/chain.yml:290]: Add optional parameters of the tipset from which to start (default the implicit head), and the height at which to stop (default 0).

**Review by anorth — COMMENTED** [review]
@eingenito I've reviewed down to `chain.yml`. I initially started pointing to the spec in too much detail, but we should just wait a couple of weeks before sweating spec details too much.

I think when you push to `master`, this PR will automatically update. I'll continue review at some point in any case.

File: api/chain.yml (hunk 15 above)
Comment (eingenito) [reply api/chain.yml:243]: It's obviously not impossible for the calling code to have already signed the message, but it's not exactly simple. Maybe it would be appropriate to specify a different content type that takes a base64 encoded binary representation of a message for signed messages and reserve the JSON content type for unsigned messages?

**Comment by eingenito** [comment]
@anorth thanks very much for the detailed review. I understand and am incorporating all the feedback.
